apps/aechannel/src/aesc_fsm_min_depth_watcher.erl

```diff
@@ -13,46 +14,202 @@
+                    andalso is_binary(element(2, element(1, X)))
+                  )
+          andalso ( is_map(element(2, X))
+                    andalso (map_size(element(2, X)) =:= 3
+                             orelse map_size(element(2, X)) =:= 4)
+                  )
+        ) ).
+-define(IS_SCENARIO(S),
+        ( (S =:= top)
+          orelse (S =:= next_block)
+          orelse (S =:= fork_switch)
+          orelse ( is_tuple(S)
+                   andalso (tuple_size(S) =:= 2)
+                   andalso (element(1, S) =:= has_tx)
+                   andalso ?IS_INFO_OF_SCENARIO_HAS_TX(element(2, S))
+                 )
+        ) ).
 
 -record(st, { parent
             , chan_id
+            , chan_vsn
+            , last_block                  %% last time we updated channel vsn
+            , last_top                    %% the block hash of the last check
+            , tx_log = aesc_window:new() :: tx_log()
+            , rpt_log = aesc_window:new() :: rpt_log()
             , closing = false
             , requests = [] }).
 
+-record(tx_log_entry,
+        { key   :: tx_log_entry_key()
+        , value :: tx_log_entry_value()
+        }).
+
+-record(rpt_log_entry,
+        { key   :: {changed_on_chain | closing_on_chain | closed_on_chain, tx_hash()}
+        , value :: #{ atom() := term() }
+        }).
+
+-type mode()       :: close
+                    | unlock
+                    | tx_hash
+                    | watch.
+
+-type req_info() :: #{ type         := any()
+                     , parent       := pid()
+                     , callback_mod := module() }.
+
+-type close_req() :: #{ mode      := close
+                      , min_depth := aec_blocks:height()
+                      , info      := req_info() }.
+-type unlock_req() :: #{ mode := unlock
+                       , info := req_info() }.
+-type watch_req() :: #{ mode := watch
+                      , info := req_info() }.
+-type tx_req() :: #{ mode      := tx_hash
+                   , tx_hash   := tx_hash()
+                   , min_depth := aec_blocks:height()
+                   , info      := req_info() }.
+
+-type req() :: close_req() | unlock_req() | watch_req() | tx_req().
+
+-type block_hash()  :: binary().
+-type tx_hash()     :: binary().
+-type tx_location() :: {block_hash(), aetx:tx_type(), aetx_sign:signed_tx()}.
+-type info_of_scenario_has_tx() :: {tx_log_entry_key(), tx_log_entry_value()}.
+-type scenario()   :: top
+                    | next_block
+                    | fork_switch
+                    | {has_tx, info_of_scenario_has_tx()}.
+-type chan_vsn()   :: undefined | { aesc_channels:round()
+                                  , aesc_channels:solo_round()
+                                  , aesc_channels:is_active()
+                                  , aesc_channels:locked_until()
+                                  , aesc_channels:state_hash() }.
+-type ch_status() :: undefined
```

Comment:
I do not see this `undefined` used.

Reply:
It would be the return value of `get_basic_ch_status_/3` before the channel object has been created. Ideally, this should never happen, since status isn't checked until the channel is active (I think).